### ISSUE TITLE
Typescript Index Signature extension methods

### ISSUE THREAD

**Issue by WayneBarnard** [Microsoft/TypeScript#19828]
I`m trying to create my own Typescript dictionary like class, but with some custom methods to use in my code.

I can create the basic dictionary like this:

```
export interface IDictionary<T> {
 [property: string]: T;
};

export class Dictionary<T> implements IDictionary<T> {
 [property: string]: T;
 constructor() {}
}
```
and this works great...but... if i try to create a extension method, such as

```
export class Dictionary<T> implements IDictionary<T> {
 [property: string]: T;
 constructor() {}
}    
public getValues(): T[] {
    return Object.values(this);
}
```
I get this error: Property 'getValues' of type '() => T[]' is not assignable to string index type 'T'.

Is this even possible? If it is, how can do I create extension methods for my this?

**Comment by rozzzly** [comment]
It is possible. It's a bit ugly. I'll show you how after I take a nap lmao

**Comment by WayneBarnard** [comment]
mmm, thats a long nap :D

**Comment by rozzzly** [comment]
```typescript
export interface MyBaseDictionary<T> {
    [key: string]: T;
}

export interface MyDictionaryInstance<T> {
    ‎getValues(): T[];
    ‎// ...
}

export type MyDictonary<T> = (
    ‎& MyBaseDictionary<T>
    ‎& MyDictionaryInstance<T>
);

export function MakeDictonary<T>(): MyDictonary<T> {
    // Create empty dict, keep reference so instance methods can reference that data--because we are not using this.
    const dict: MyDictonary<T> = {} as any;
    ‎// Add dict's instance methods to the empty dict via `Object.defineProperty` so they're not enumerable via `Object.values`
    ‎Object.defineProperty(dict, 'getValued', { 
        enumerable: false,
        ‎configurable: false,
        ‎writable: false,
        ‎value: () => {
    ‎.      return Object.getValues (dict);
        }
    });
    ‎return dict;
}
```

**Comment by rozzzly** [comment]
Okay, so super rough and untested ( typed this up on my phone ) but hopefully it gives you an idea what I was getting at. It's not an es2105 class, but it's instances are quite similar to what it looks like what you are trying to achieve. Lemme know if you have questions.

PS: personally, I'd suggest redoing your API. I'm assuming you posted a somewhat contrived example and that end result you're invisioning is a lot like something I have toyed arround with in the past. I ended up ditching the pattern shown above, having the values nested in a property at a depth equal to those instance methods. Just wasn't worth the hassle.

But that was for my usecase, you'll have to weigh  the pros/cons of such a pattern for yourself.
